### PR TITLE
UPBGE: Add an option to create objects list looping in c++

### PR DESCRIPTION
**Pull request by youle31**
temp work. creates objects list that match name or property. Keyword args

test file: http://pasteall.org/blend/index.php?id=44729

requires performances tests

Performance test file: http://pasteall.org/blend/index.php?id=44738 ... Wait... I fix a mistake in the code

So not big difference on a list of 100 objects. This only makes difference for really big lists. But before sending it to trash, maybe we could test with getMatching(materialName = "...") or with other things... Also on other computers than mine who has a powerful CPU

### CODE REVIEW

**Comment by DCubix** [comment]
283fps = with getMatching
260fps = without
Seems like a good job to me :+1:

**Comment by panzergame** [comment]
@youle31 : The function you added check for name and property, these two attributes are shared by all CValue sub class and CValue is contained in CListValue. Why not move this function to CListValue ?

File: source/gameengine/Ketsji/KX_Scene.cpp [hunk 1]
```diff
@@ -2533,4 +2534,33 @@ KX_PYMETHODDEF_DOC(KX_Scene, get, "")
 	return def;
 }
 
+/* Find gameobjects matching a name, a property... */
+KX_PYMETHODDEF_DOC(KX_Scene, getMatching, "scene.getMatching(objectsName, objectsProperty...)")
+{
+	PyObject *objectslist = PyList_New(0);
+	char *objectsName = "";
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_Scene.cpp]: You could use NULL to remove strlen check later.

File: source/gameengine/Ketsji/KX_Scene.cpp [hunk 2]
```diff
@@ -2533,4 +2534,33 @@ KX_PYMETHODDEF_DOC(KX_Scene, get, "")
...
+	char *objectsProperty = "";
+
+	static const char *kwlist[] = { "name", "property", NULL };
+
+	if (!PyArg_ParseTupleAndKeywords(args, kwds, "|ss:getMatching", (char **)(kwlist), &objectsName, &objectsProperty))
+		return NULL;
+
+	for (CListValue::iterator<KX_GameObject> it = m_objectlist->GetBegin(), end = m_objectlist->GetEnd(); it != end; ++it) {
+		KX_GameObject *gameobj = *it;
+		if (strlen(objectsName)) {
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_Scene.cpp]: The user want to have a list of object with giving name AND property. or a list of object with the giving name OR property ?

File: source/gameengine/Ketsji/KX_Scene.cpp [hunk 3]
```diff
@@ -2533,4 +2534,33 @@ KX_PYMETHODDEF_DOC(KX_Scene, get, "")
 	return def;
 }
 
+/* Find gameobjects matching a name, a property... */
+KX_PYMETHODDEF_DOC(KX_Scene, getMatching, "scene.getMatching(objectsName, objectsProperty...)")
+{
+	PyObject *objectslist = PyList_New(0);
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_Scene.cpp]: Why not return a CListValue with python owner proxy ?

**Comment by panzergame** [comment]
I propose you to use an optional regex system. For example if the user want to look at all the object starting with name "Cube" and containing the property "Prop":
```
scene.objects.getMatching("^Cube.*", "Prop")
```

Regex could be used with the C++11 regex library: http://en.cppreference.com/w/cpp/regex.

**Comment by youle31** [comment]
Yes. About regular expressions, why not.
About this patch, I just begun to implement it to see if this could be increase performances when we create lists of objects. But for now, I don't know if this is really useful. For 100 objects, on my computer I see no performance difference beetween a list created with comprehension lists or with the actual getMatching. For 1000 objects we can see a diference but not as important as expected. So I don't know if I continue to work on it or not.

@DCubix : What do you think? What could be the usages? Do you think it's still interesting? Would you want to continue to work on it?

@panzergame && @DCubix : Do you think we could use this patch to: instead of just create a list of objects, we ask to apply a function/method for a list of objects: e.g: scene.getMatching(property = "prop", function = "applyRotation((0.0, 0.0, 0.1) False")) ?

**Comment by lordloki** [comment]
For the perfomance test file I obtain 55fps using getMatching and 41fps using current way.

**Comment by BluePrintRandom** [comment]
I need to batch update terrain mesh instances using a table of stored
vertex offsets,  normals and vertex color for splat map blending.

(so as the actor moves forward new patches are loaded and deformed and
textured and old patches behind you unload)

I will also need batch add object if possible

scene.addBatch([ ['buildingpiece', loc xyz, rot], ['buildingPiece2', loc
xyz, rot]])

**Comment by youle31** [comment]
Thanks @lordloki for performances test. On your computer it seems it makes a big difference, so we could eventually continue to work on this patch.

**Comment by youle31** [comment]
panzergame commited another approach with regex here https://github.com/UPBGE/blender/commit/4441d58bc3c051f61989761453650470b2245418 so I close this

**Pull request closed without merging**